[ConformanceLookup] Don't allow skipping inherited unavailable conformances in favor of explicit available ones.

hborla opened the pull request:
The type checker does not support the notion of multiple protocol conformances; there can only be one conformance, and if that conformance is unavailable, you cannot specify your own available conformance. This is important for Sendable checking; if a framework specifies that a type is explicitly not Sendable with an unavailable Sendable conformance, clients cannot ignore Sendable violations involving that type. If a superclass wants to allow subclasses to add a Sendable conformance, it should not declare an unavailable Sendable conformance.

Note that this change still does not warn if the original unavailable conformance is declared with `@_nonSendable` due to the way `@_nonSendable` expansion happens during implicit `Sendable` derivation, after no explicit conformance has been found.

Resolves: rdar://124423524

hborla committed:
[ConformanceLookup] Don't allow skipping inherited unavailable conformances
in favor of explicit available ones.

The type checker does not support the notion of multiple protocol
conformances; there can only be one conformance, and if that conformance
is unavailable, you cannot specify your own available conformance. This
is important for Sendable checking; if a framework specifies that a type
is explicitly not Sendable with an unavailable Sendable conformance,
clients cannot ignore Sendable violations involving that type. If a
superclass wants to allow subclasses to add a Sendable conformance, it
should not declare an unavailable Sendable conformance.

hborla committed:
[ConformanceLookup] Always prefer unavailable Sendable conformances from the
defining module, and diagnose redundant Sendable conformances.

We still allow re-stating inherited unchecked Sendable conformances in
subclasses because inherited Sendable conformances are surprising when
they opt out of static checking. Otherwise, warning on redundant Sendable
conformances nudges programmers toward cleaning up unnecessary retroactive
Sendable conformances over time as libraries incrementally add the
conformances directly.

ktoso reviewed (approved):
Tricky impl hah, looks okey

hborla commented on test/Concurrency/redundant_sendable_conformance.swift:
```diff
@@ -0,0 +1,19 @@
+// expected-warning@-1 {{conformance of 'NonSendableClass' to protocol 'Sendable' was already stated in the type's module 'SendableConformances'}}
+
+typealias CheckNonSendableClass = RequireSendable<NonSendableClass>
+// expected-error@-1 {{conformance of 'NonSendableClass' to 'Sendable' is unavailable}}
```
This makes it impossible for libraries to stage in unavailable Sendable conformances when they have clients that have adopted `-swift-version 6`. This needs to either be a warning or we need to only warn on the redundant conformance while still picking that conformance as the best one, so that errors only appear when the retroactive conformance is removed.

hborla committed:
[ConformanceLookup] Just kidding, the compiler needs to prefer available
Sendable conformances for source compatibility.

If conformance lookup always prefers the conformance from the defining module,
libraries introducing unavailable Sendable conformances can break source in
clients that declare retroactive Sendable conformances. Instead, still prefer
the available conformance, and always diagnose the client conformance as
redundant (as a warning). Then, when the retroactive conformance is removed,
the errors will surface, so the programmer has to take explicit action to
experience the source break.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla committed:
[ProtocolConformance] Compute the correct un-availability for inherited
conformances.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla commented:
The macOS smoke test is failing due to a driver test failure unrelated to this change. Looks like a revert is being tested at https://github.com/swiftlang/swift-driver/pull/1662

hborla commented:
@swift-ci please smoke test macOS
